language/vm/src/views.rs

```diff
@@ -509,6 +509,58 @@ impl<'a, T: ModuleAccess> LocalsSignatureView<'a, T> {
     }
 }
 
+/// Returns the kind of the signature token in the given context (module, function/struct).
+/// The context is needed to determine the kinds of structs & type variables.
+// TODO: rework views to make the interfaces here cleaner.
+pub fn kind_impl(
+    (struct_handles, type_formals): (&[StructHandle], &[Kind]),
+    ty: &SignatureToken,
+) -> Kind {
+    use SignatureToken::*;
+
+    match ty {
+        // These primitive types have kind unrestricted.
+        Bool | U64 | String | ByteArray | Address | Reference(_) | MutableReference(_) => {
+            Kind::Unrestricted
+        }
+
+        // To get the kind of a type parameter, we lookup its definition in the formals.
+        TypeParameter(idx) => type_formals[*idx as usize],
+
+        Struct(idx, tys) => {
+            // Get the struct handle at idx. Note the index could be out of bounds.
+            let sh = &struct_handles[idx.0 as usize];
+
+            if sh.is_nominal_resource {
+                return Kind::Resource;
+            }
+
+            // Gather the kinds of the type actuals.
+            let kinds = tys
+                .iter()
+                .map(|ty| kind_impl((struct_handles, type_formals), ty))
+                .collect::<Vec<_>>();
+
+            // Derive the kind of the struct.
+            //   - If any of the type actuals has kind `all`, then the struct has kind `all`.
+            //     - `all` means some part of the type can be either `resource` or `unrestricted`.
+            //     - Therefore it is also impossible to determine the kind of the type as a whole,
+            //       and thus `all`.
+            //   - If none of the type actuals has kind `all`, then the struct is a resource if and
+            //     only if one of the type actuals has kind `resource`.
+            kinds
+                .iter()
+                .fold(Kind::Unrestricted, |acc_kind, next_kind| {
+                    match (acc_kind, next_kind) {
```

Comment:
It would be good to add this nice general-purpose logic as something like `pub fn join(&self, kind: &Kind) -> Kind` in `impl Kind`.